### DISCUSSION TITLE
Adding Prefixes to Generated SubModel Names

### DISCUSSION THREAD

**Discussion by killik-dan-stubbs** [hiranthaR/Json-to-Dart-Model#108] [Ideas]
# Adding Prefixes to Generated SubModel Names

Over the past year or so, I've really enjoyed using Json to Dart Model for generating freezed classes from json responses that I get from my company's backend. It saves so much time and effort; I'm very grateful for the work that has gone into it.

## Issue: Confusing Model Names

I have one issue with class generation though which keeps with extension from perfection. The problem is that there is no way, as far as I can tell, to assign a standard prefix to a generated model. 

The reason why this is important is that you could generate two separate models for different api calls which both have a 'contact' map field. After generation, both of these models would be called `Contact`. If looking using devtools and trying to find one contact over another, it can be hard to differentiate which model is the correct to use.

If we had the ability to add a Prefix to all generated models, we could have models like `RouteOneContact` and `RouteTwoContact` which would be much easier to read. 

Below is the response of an imaginary api call which returns a users Contact information.

```json
{
  "contact": {
    "name": "John Doe",
    "email": "john.doe@test.com",
    "phone": "123-456-7890"
  },
  "address": {
    "street": "123 Main St",
    "city": "Anytown",
    "state": "NY",
    "zip": "12345"
  }
}
```

Currently, the generated model would look like this with the Base Class name set to `UserDetails`:

```dart
class UserDetails {
  UserDetails({this.contact, this.address});

  Contact? contact;
  Address? address;
}

/// As you can tell, the Contact class could be confused 
/// with another Contact class
class Contact {
  Contact({this.name, this.email, this.phone});

  String? name;
  String? email;
  String? phone;
}

class Address {
  Address({this.street, this.city, this.state, this.zip});

  String? street;
  String? city;
  String? state;
  String? zip;
}
```

## Proposed Solution 1: Prefix Input

Perhaps an extra question could be added to the flow of the json to dart model generation which would allow the user to add a prefix to the generated model. This would be a simple string input that would be added to the model name.

### Example

```
Q: What prefix would you like to add to the generated model?

> UserDetails
```

This would then generate the following models:

```dart
/// Please note that the Base Class Name would not be effected
/// by the prefix input
class UserDetails {
  UserDetails({this.contact, this.address});

  UserDetailsContact? contact;
  UserDetailsAddress? address;
}

/// As you can tell, the Contact class could be confused 
/// with another Contact class
class UserDetailsContact {
  UserDetailsContact({this.name, this.email, this.phone});

  String? name;
  String? email;
  String? phone;
}

class UserDetailsAddress {
  UserDetailsAddress({
    this.street, 
    this.city, 
    this.state, 
    this.zip
  });

  String? street;
  String? city;
  String? state;
  String? zip;
}
```

The fact that the base class name would not be effected by the prefix input is not ideal. Perhaps the base class name could be used as a prefix if the user does not input one.

## Proposed Solution 2: Base Class name as Prefix of Sub Models

Another solution could be to ask the user if the base class name should be used as a prefix for all sub models. This would be a simple boolean input which could either be asked or added within a config file.

As a side note, the file names for the models may not need to be changed with the prefix but it may be a good idea to add them anyway for consistency and to avoid potential linting errors.

### Example

```
Q: Use Base Class Name as a prefix for all sub models? (Yes/No)

> Yes
```

This would then generate the following models:

```dart
/// Please note that the Base Class Name would not be effected
class UserDetails {
  UserDetails({this.contact, this.address});

  UserDetailsContact? contact;
  UserDetailsAddress? address;
}

/// As you can tell, the Contact class is now clearly to be used
/// with the UserDetails class
class UserDetailsContact {
  UserDetailsContact({this.name, this.email, this.phone});

  String? name;
  String? email;
  String? phone;
}

class UserDetailsAddress {
  UserDetailsAddress({this.street, this.city, this.state, this.zip});

  String? street;
  String? city;
  String? state;
  String? zip;
}
```

## Finishing Up

What do you all think? Would this be a useful feature to add to the json to dart model generation process? If I get a chance, I may make a fork of the project and try to implement this feature myself.

**Comment by iamarnas** [comment]
@killik-dan-stubbs Hi 👋
Thank you for your such detailed information. This can be done by adding the parent name as prefix. Just for the bigger Json file should be needed analyzer for the dublicate names. However Json To Dart Model generator do not handle dublicates, that why here no option generate to single file. But by adding prompt to ask for the parent name as prefix it is possible. I'll keep it in mind, in my plans is to rewrite the generator core for more flexibility with the better analyser and better code readability. That why is no updates at the moment.
